homeassistant/components/mqtt/vacuum/__init__.py

```diff
@@ -27,26 +30,35 @@
 )
 
 
-def validate_mqtt_vacuum_discovery(value):
+def validate_mqtt_vacuum_discovery(config_value: ConfigType) -> ConfigType:
     """Validate MQTT vacuum schema."""
-    schemas = {LEGACY: DISCOVERY_SCHEMA_LEGACY, STATE: DISCOVERY_SCHEMA_STATE}
-    return schemas[value[CONF_SCHEMA]](value)
+    schemas: dict[str, vol.Schema] = {
+        LEGACY: DISCOVERY_SCHEMA_LEGACY,
+        STATE: DISCOVERY_SCHEMA_STATE,
+    }
+    config: ConfigType = schemas[config_value[CONF_SCHEMA]](config_value)
+    return config
 
 
 # Configuring MQTT Vacuums under the vacuum platform key is deprecated in HA Core 2022.6
-def validate_mqtt_vacuum(value):
+def validate_mqtt_vacuum(config_value: ConfigType) -> ConfigType:
     """Validate MQTT vacuum schema (deprecated)."""
-    schemas = {LEGACY: PLATFORM_SCHEMA_LEGACY, STATE: PLATFORM_SCHEMA_STATE}
-    return schemas[value[CONF_SCHEMA]](value)
+    schemas: dict[str, vol.Schema] = {
+        LEGACY: PLATFORM_SCHEMA_LEGACY,
+        STATE: PLATFORM_SCHEMA_STATE,
+    }
```

Comment:
Hm – `PLATFORM_SCHEMA_LEGACY` seems to be a `vol.All`, not a `vol.Schema`...

Reply:
Technically, `Schema` and `All` are not part of the same inheritance tree so this should indeed be resolved.
It isn't picked up automatically by `mypy` because `voluptuous` isn't typed.

If this is a common pattern, maybe it would make sense to add a type alias to `config_validation` helper:
```python
ConfigTypeValidator = Callable[[Any], ConfigType]
```

Reply:
Looking again, I think the easiest is to let it be inferred.
It is not needed.
```suggestion
    schemas = {LEGACY: PLATFORM_SCHEMA_LEGACY, STATE: PLATFORM_SCHEMA_STATE}
```